ogcapi-records part2 - Facets

davidblasby opened the pull request:
# PR Title

This is an implementation of the OGCAPI-Records Part 2 (Facets) specification as well as support for (limited) CQL expressions.



## Description

The basic idea is to similar to how the queryables are implemented.    Will allow some type of configuration to turn the facets on/off.  Will look at common ones used by the GN4 search user interface.

This will be very simple (not the advanced) facets.

1. OpenAPI (Swagger) definitions

-  For the `/<collectonID>/facets` endpoint

cf, https://docs.ogc.org/DRAFTS/25-013.html

## Type of Change
Select one:
- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Refactoring

## Related Issue
Closes/updates # (optional):

## Approach
<!-- Outline your solution and any key decisions. Bullet points are fine. -->

## How to Verify
<!-- Steps for reviewer to reproduce and verify your change. -->
1. 
3. 
4. 

## Checklist
- [ ] Code is formatted and linted
- [ ] All tests pass locally and in CI
- [ ] Tests updated (if applicable)
- [ ] Documentation updated (if applicable)

<!--
Optional: add any extra notes, credits (e.g., "Funded by ..."), or dependencies.
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=100) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=100) before we can accept your contribution.<br/>**3** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: david-blasby<br/>:white_check_mark: PascalLike<br/>:white_check_mark: juanluisrp<br/>:x: josegar74<br/>:x: fxprunayre<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/geonetwork/geonetwork?pullRequest=100) it.</sub>

davidblasby commented:
hmmm - a lot of commits here.  Likely a rebase on main (or similar).

davidblasby commented:
I've added in the GN-UI test data and created the start of some integration tests - It will spin-up (`testcontainers`) a postgresql and elastic instance,  both pre-filled with some data.

You can see some trivial tests in `TrivalRecordsMvcTests` that use this (it spins up a GN5 instance and executes a few ogcapi-records requests against the pre-configured postgres/elastic).

Working well.  The main reason I added it is because there are some "weird" ways that elastic responds to seemingly equivalent queries, so I want to put in tests for the facets and CQL queries (and likely some of the other queries as well).

PascalLike commented:
Thanks @davidblasby for this! This looks ready for a review.

I'll do some local testing in the next few days. This is one of the first implementations of [OGC API Record Part 2](https://docs.ogc.org/DRAFTS/25-013.html). 

I've also invited the maintainers: it would be great to have more feedback. 

When it's merged, we should update [this](https://github.com/opengeospatial/ogcapi-records/blob/master/implementations.md#geonetwork-opensource)

PascalLike reviewed (changes requested):
@davidblasby, it's evident there's a lot of work in this PR, congratulations! I like the solutions you've implemented; the code is very smart and clear. 
I've added some comments inline; some are minor, others are simple questions.
From what I can tell by comparing them to the specifications, we're very close, if not yet aligned, with the state of the art with OGC API Records Part 2. I'd say go all out and add the [conformance classes](https://docs.ogc.org/DRAFTS/25-013.html#_conformance) to the /conformance endpoint.

Before merging, I'd also ask you to rebase the main branch, where there are some changes. I've already run several non-regression tests (all good!), but I prefer to repeat them once everything is aligned.

When you think everything's ready, I'll run one last test, and then I think it's okay to merge :+1:

davidblasby commented on config/application.yml:
```diff
@@ -16,6 +16,37 @@ geonetwork:
     ogcapi-records-json:
       includeRawXML: false
       includeElasticJson: false
+    ogcapi-property-mapping:
```
I'm working on this right now - I doubt it will stay in the application.yaml.  I expect it will move to the database so it will be user-changeable (but require a restart).

Its just here for the moment.

davidblasby commented on src/modules/ogcapi-records/ogcapi-records-integration-tests/pom.xml:
```diff
@@ -0,0 +1,163 @@
+    <dependency>
+      <groupId>org.springdoc</groupId>
+      <artifactId>springdoc-openapi-starter-common</artifactId>
+      <version>2.6.0</version>
```
moved - I had copied this from the `src/apps/geonetwork/pom.xml`

davidblasby commented on src/modules/ogcapi-records/ogcapi-records-openapi-autogen/src/ogc-openapi-schema/openapi/ogcapi-records-1-example-ref-schema-repo.yaml:
Yes - most of this was taken from the ogc's repo.  Its been modified to produce better java code.  

There are a few minor changes - mostly around where the spec allows multiple types.  For example, an `id` can be either a `number` or a `string` in the spec.  We accept (incoming) either numbers or string, but we always output a string (which is consistent with how GN handles ids).

More info about the modifications are here:
`src/modules/ogcapi-records/ogcapi-records-openapi-autogen/src/ogc-openapi-schema/README.md`

davidblasby commented on src/shared/ogcapi-records/src/main/java/org/geonetwork/ogcapi/service/facets/FacetsResponseInjector.java:
```diff
@@ -0,0 +1,395 @@
+            var b = new OgcApiRecordsFacetResultBucketDto();
+            b.setCount((int) elasticBucket.docCount());
+
+            b.setMin(elasticBucket.keyAsString());
```
removed - good catch!

davidblasby commented on src/shared/ogcapi-records/src/main/java/org/geonetwork/ogcapi/service/facets/FacetsResponseInjector.java:
```diff
@@ -0,0 +1,395 @@
+@Component
+public class FacetsResponseInjector {
+
+    @Autowired
```
The lomboc @AllArgsConstructor makes this pretty easy to change.

davidblasby commented on src/shared/ogcapi-records/src/main/java/org/geonetwork/ogcapi/service/facets/RecordsFacetsElasticQueryBuilder.java:
```diff
@@ -0,0 +1,176 @@
+        var aggregations = new HashMap<String, Aggregation>();
+
+        var defaultBucketCount =
+                facets.getDefaultBucketCount() == null ? Integer.MAX_VALUE : facets.getDefaultBucketCount();
```
Added FacetsResponseInjector.DEFAULT_MAX_BUCKETS

davidblasby commented:
rebased based on `origin/main`
